### PR TITLE
[REF] web_editor, *: make SnippetsMenu an OWL component

### PR DESCRIPTION
**Pull request by detrouxdev**
The goal of this PR is to convert the SnippetsMenu from a "legacy" Widget to an OWL Component. We start with **only** the SnippetsMenu and not SnippetEditor, SnippetOption, or UserValueWidget.

We also apply the OWL way of doing things. This means that SnipptesMenu's instances are no longer accessible from the `wysiwyg` instance. Instead, a bus is introduced for each method that could not be adapted into a prop.

Because the SnippetsMenu template is now rendered by OWL, we remove its UI XML code from the server view (`web_editor.snippets`), and move it into a static frontend XML Template (`web_editor.SnippetsMenu`). Snippets are still defined in the old backend view, but the template has been heavily simplified.

task-3323657

Enterprise PR: https://github.com/odoo/enterprise/pull/55980
Upgrade PR: https://github.com/odoo/upgrade/pull/5681

### CODE REVIEW

**Comment by qsm-odoo** [comment]
Marking this as ready for review. For info:
- Second commit of this PR will be gone, it is being done at https://github.com/odoo/odoo/pull/156658
- Already gone though (multiple) india testing. I'll test intensively by myself now but if someone wants to have a go, it's not too much :wink: 
- Plan is for me to start reviewing, then merging as soon as possible (this week?) once the freeze of 17.2 has been made.

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
@detrouxdev Here is a first review. There are quite some superficial comments but also questions about some more generic things I would change. Tell me what you think :+1: 

Thanks for the work :+1: Tell me if https://github.com/odoo/odoo/pull/156658 is ready then come back to this in priority (forget about the reviews of the week for a few days if necessary). Also, there are rebase conflicts here, could you rebase? :pray: (try to always do it without any change from now so that I can track more easily the changes you make here).

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 1]
```diff
@@ -2328,8 +2345,12 @@ export class Wysiwyg extends Component {
             },
         }
     }
-    _insertSnippetMenu() {
-        return this.snippetsMenu.insertBefore(this.$el);
+    async _insertSnippetMenu() {
+        const snippetsMenuMountedProm = new Deferred();
```
[the remark this answers is not in the document]
Comment (detrouxdev) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:2349]: This is not from jQuery, but from our own `@web/core/utils/concurrency` 😄 but I'll try to keep it in mind as I fix things 👍🏻 

Though speaking of this, what do you think about it? It's to replace the fact that OWL does not directly offer a way to await for a component to be mounted. I do not like this as it means the parent component is stuck waiting for its child, which I think is against OWL principles. I was just scared of the possible race conditions I would introduce if I try to re-arrange the code from the Wysiwyg class.

In any case, I'll add a TODO next to the line cause improvements could certainly be made to remove this type of code.

[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:2349]: > In any case, I'll add a TODO next to the line cause improvements could certainly be made to remove this type of code.

That's what I think about it :wink: Should be removed but can be kept for now :)

> This is not from jQuery, but from our own @web/core/utils/concurrency

Oh, did not know they made that... why not :)

File: addons/mass_mailing/static/src/js/mass_mailing_html_field.js [hunk 2]
```diff
@@ -190,14 +201,14 @@ export class MassMailingHtmlField extends HtmlField {
 
     async _onSnippetsLoaded() {
         if (status(this) === 'destroyed') return;
-        if (this.wysiwyg.snippetsMenu && $(window.top.document).find('.o_mass_mailing_form_full_width')[0]) {
+        if (this.wysiwyg.state.showSnippetsMenu && $(window.top.document).find('.o_mass_mailing_form_full_width')[0]) {
```
[the remark this answers is not in the document]
Comment (detrouxdev) [reply addons/mass_mailing/static/src/js/mass_mailing_html_field.js:204]: Turns out, this code is never executed anymore as the class `.o_mass_mailing_form_full_width` was removed by [this commit](https://github.com/odoo/odoo/commit/5226dc972560870d03e7e4bc5a838f1951b90ac7). This broke scrolling for snippets that are already in the page.
Should I make a fix for it first? I could potentially remove these lines with the fix too.

[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_html_field.js:204]: > Should I make a fix for it first?

I guess? I did not understand what it is about. Either fix what's to be fixed or adapt so the fix can be solved later by the related team? :shrug: Whatever allows us to merge this as soon as possible :+1:

**Comment by detrouxdev** [comment]
I went through a first part of comments, I tried to leave biggest changes in separate commits so that it's easier to see them.

I have a few more changes in stash but I will first handle the rebase on Monday.

File: addons/mass_mailing/__manifest__.py [hunk 3]
```diff
@@ -150,8 +150,10 @@
             'mass_mailing/static/src/snippets/s_showcase/options.js',
             'mass_mailing/static/src/snippets/s_rating/options.js'
         ],
-        'web_editor.assets_legacy_wysiwyg': [
+        'mass_mailing.assets_legacy_wysiwyg': [
```
[the remark this answers is not in the document]
Comment (detrouxdev) [reply addons/mass_mailing/__manifest__.py]: I added a commit that renames the bundles in which I have a little bit of documentation of why I'm doing so, then I made a separate commit in which I create the bundle, explaining why I'm doing so. It's not documentation directly on the bundles, I can do that too if you want 👍🏻

File: addons/mass_mailing/static/src/js/mass_mailing_html_field.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (detrouxdev) [reply addons/mass_mailing/static/src/js/mass_mailing_html_field.js:214]: I kept this line as is for now, but I changed the next lines to be a local state. Meaning that if the test on snippetsMenu was done so that we don't access it when it is undefined, it should be okay now to remove that part from the `if` as we can access the local state regardless.
[I asked the editor team for some clarification](https://discord.com/channels/678381219515465750/687338152372994236/1222110710595846174) before moving one though.

File: addons/mass_mailing/static/src/js/mass_mailing_wysiwyg.js [hunk 4]
```diff
@@ -31,9 +31,9 @@ export class MassMailingWysiwyg extends Wysiwyg {
         super.toggleLinkTools({
             ...options,
             // Always open the dialog when the sidebar is folded.
-            forceDialog: options.forceDialog || this.snippetsMenu.folded
+            forceDialog: options.forceDialog || this.snippetsMenuProps.folded
         });
-        if (this.snippetsMenu.folded) {
+        if (this.snippetsMenuProps.folded) {
```
[the remark this answers is not in the document]
Comment (detrouxdev) [reply addons/mass_mailing/static/src/js/mass_mailing_wysiwyg.js]: I changed it to props and individual entries in the state, you're right it was cleaner / better 👍🏻

File: addons/mass_mailing/static/src/js/snippets.editor.js [hunk 5]
```diff
@@ -71,24 +80,24 @@ export const MassMailingSnippetsMenu = snippetsEditor.SnippetsMenu.extend({
...
+    }
     /**
      * @override
      */
-    _updateRightPanelContent: function ({content, tab}) {
-        this._super(...arguments);
-        this.$('.o_we_customize_design_btn').toggleClass('active', tab === this.tabs.DESIGN);
-    },
+    _patchForComputeSnippetTemplates($html) {
+        this.env.switchImages(this.themeParams, $html);
```
[the remark this answers is not in the document]
Comment (detrouxdev) [reply addons/mass_mailing/static/src/js/snippets.editor.js]: The env is [the OWL way to share things across components](https://github.com/odoo/owl/blob/master/doc/reference/environment.md) when not all components have to use it:
Say component tree  A => B => C, A (parent) wants to share something with C (child), then it can use `useSubEnv/useChildSubEnv` and in case it needs to be reactive, use reactive or useState and put that reactive variable in the env.

I put some methods here for that purpose, but I could also use `options` since that variable is passed from parent to child.

[I did end up adding some reactive state](https://github.com/odoo/odoo/blob/d6078dbbe8864bc911cfbc7b6feccebbe28be9cc/addons/mass_mailing/static/src/js/mass_mailing_html_field.js#L31) in there to remove the bus trigger as well.

File: addons/mass_mailing/static/src/xml/mass_mailing.editor.xml [hunk 6]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="utf-8"?>
+<templates id="template" xml:space="preserve">
+    <t t-name="mass_mailing.SnippetsMenu" t-inherit="web_editor.SnippetsMenu" t-inherit-mode="primary">
```
[the remark this answers is not in the document]
Comment (detrouxdev) [reply addons/mass_mailing/static/src/xml/mass_mailing.editor.xml:3]: I was following how it was done in the files that were in the same folder. Actually, it seems like a lot of XML files which contains OWL component are done this way too:
- [autocomplete.xml](https://github.com/odoo/odoo/blob/master/addons/web/static/src/core/autocomplete/autocomplete.xml#L21) 
- [datetime_pick.xml](https://github.com/odoo/odoo/blob/master/addons/web/static/src/core/datetime/datetime_picker.xml)
- [popover.xml](https://github.com/odoo/odoo/blob/master/addons/web/static/src/core/popover/popover.xml)
- ...

There's more and there's some which have the other way too, but I thought it'd probably be better to match the same style in the module. Should I still do it? What's the advantage?

File: addons/web_editor/static/src/xml/snippets.xml [hunk 7]
```diff
@@ -133,5 +133,128 @@
...
+    <t t-name="web_editor.SnippetsMenu">
+        <div id="oe_snippets" t-ref="snippets-menu" t-on-mousedown="_onMouseDown">
+            <div class="o_we_website_top_actions">
+                <div class="o_we_external_history_buttons d-flex">
+                    <button type="button" t-on-click="_onUndo" class="btn btn-secondary fa fa-undo" t-att-disabled="!state.canUndo"/>
+                    <button type="button" t-on-click="_onRedo" class="btn btn-secondary fa fa-repeat" t-att-disabled="!state.canRedo"/>
+                </div>
+                <form class="ms-auto d-flex">
+                    <!-- Uncomment the following line when the mobile preview will be available. -->
+                    <!-- <button type="button" class="btn btn-secondary fa fa-mobile" name="mobile" data-action="mobilePreview"/> -->
```
[the remark this answers is not in the document]
Comment (detrouxdev) [reply addons/web_editor/static/src/xml/snippets.xml]: It was moved in the view, I just moved it into the frontend template with the rest. I did not remove it, but it's a good time to do so 👍🏻

File: addons/website/static/src/xml/web_editor.xml [hunk 8]
```diff
@@ -37,34 +37,12 @@
...
-        <xpath expr="//div[1]" position="inside">
-            <div t-if="props.showTextHighlights" id="highlight" class="btn-group">
-                <div title="Add Text Highlight Effects"
-                    class="btn o_we_text_highlight"
-                    data-text-selector=".o_text_highlight">
-                    <span class="oi oi-text-effect oi-fw"></span>
-                    <span>Highlight</span>
-                </div>
-            </div>
-        </xpath>
```
[the remark this answers is not in the document]
Comment (detrouxdev) [reply addons/website/static/src/xml/web_editor.xml:68]: The part moved is done in a slot because it is just adding XML where the slot would be, the rest still needs to be done using XPath.

File: addons/mass_mailing/static/src/js/mass_mailing_html_field.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_html_field.js]: > A fix in stable will still need to be done to either fully remove the dead code, or adapt it.

Removing dead code in stable should not be done. If something needs to be adapted to fix things, you can explain it to the editor team, let's focus on master :+1:

File: addons/mass_mailing/static/src/xml/mass_mailing.editor.xml (hunk 6 above)
[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/mass_mailing/static/src/xml/mass_mailing.editor.xml:3]: Apparently, there are:
- 965 as you did (well, variants of it)
- 112 as I want

Unfortunately I don't think that "the majority did it that way" is really a good argument :grimacing: 

> What's the advantage?

Well... 4 more characters available for the XML your editor will show :grimacing: 

> Should I still do it?

I'll ask the framework JS team if they recommend the opposite way I asked. If they don't care, I'll prefer we do it :wink:

[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/mass_mailing/static/src/xml/mass_mailing.editor.xml:3]: They don't really care but they do as you did in general. So I let you decide :wink:

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 9]
```diff
@@ -1768,8 +1775,8 @@ export class Wysiwyg extends Component {
      * @returns {Promise}
      */
     waitForEmptyMutexAction() {
-        if (this.snippetsMenu) {
-            return this.snippetsMenu.execWithLoadingEffect(() => null, false);
+        if (this.state.showSnippetsMenu) {
+            return this.mutex.getUnlockedDef();
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:1779]: Did you adapt this to still have an effect when `waitForEmptyMutexAction` is called?

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 10]
```diff
@@ -1810,7 +1821,8 @@ export class Wysiwyg extends Component {
             onColorpaletteDropdownHide: this.onColorpaletteDropdownHide.bind(this),
             textColorPaletteProps: this.colorPalettesProps.text,
             backgroundColorPaletteProps: this.colorPalettesProps.background,
-        }
+            showRemoveFormat: !!(this.state.snippetsMenuFolded || !this.options.snippets),
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:1824]: I think the !! is useless?

File: addons/website/__manifest__.py [hunk 11]
```diff
@@ -275,14 +274,14 @@
...
+            ('include', 'web_editor.assets_snippets_menu'),
             ('include', 'website.assets_wysiwyg'),
         ],
         'website.backend_assets_all_wysiwyg': [
             ('include', 'web_editor.backend_assets_wysiwyg'),
-            ('include', 'web_editor.assets_legacy_wysiwyg'),
+            ('include', 'web_editor.assets_snippets_menu'),
             ('include', 'website.assets_wysiwyg'),
-            'website/static/src/components/wysiwyg_adapter/wysiwyg_adapter.js',
+            'website/static/src/components/wysiwyg_adapter/*',
```
Comment (qsm-odoo) [inline addons/website/__manifest__.py:284]: Sure about this? I see there is `addons/website/static/src/components/wysiwyg_adapter/page_options.js` which would now be matched but you don't see to modify it in any way.

File: addons/website/static/src/components/wysiwyg_adapter/toolbar_patch.js [hunk 12]
```diff
@@ -1,15 +0,0 @@
...
-patch(Toolbar.props, {
-    ...Toolbar.props,
-    showAnimateText: { type: Boolean, optional: true },
-    showTextHighlights: { type: Boolean, optional: true },
-});
-patch(Toolbar.defaultProps, {
-    ...Toolbar.defaultProps,
-    showAnimateText: false,
-    showTextHighlights: false,
-});
```
Comment (qsm-odoo) [inline addons/website/static/src/components/wysiwyg_adapter/toolbar_patch.js:15]: Where did that go?

File: addons/website/static/src/js/editor/snippets.editor.js [hunk 13]
```diff
@@ -313,73 +369,52 @@ const wSnippetMenu = weSnippetEditor.SnippetsMenu.extend({
...
-                ${this.$body.hasClass('o_animated_text_highlighted') ? 'fa-eye text-success' : 'fa-eye-slash'}"
-                title="${_t('Highlight Animated Text')}"
-                aria-label="Highlight Animated Text">
-            </we-button>
-        `));
-        this._toggleTextOptionsButton(".o_we_animate_text");
-        this._toggleHighlightAnimatedTextButton();
-        this._toggleTextOptionsButton(".o_we_text_highlight");
+        super._addToolbar(...arguments);
+        this.state.animatedTextHilighted = this.$body[0].classList.contains("o_animated_text_highlighted");
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.editor.js:385]: Hilighted is not a word :wink:

File: addons/website/static/src/js/editor/snippets.editor.js [hunk 14]
```diff
@@ -313,73 +369,52 @@ const wSnippetMenu = weSnippetEditor.SnippetsMenu.extend({
...
         return sel.rangeCount && [...this.getEditableArea()].some(el => el.contains(sel.anchorNode));
-    },
+    }
     /**
      * @override
      */
     _isMobile() {
-        return wUtils.isMobile(this);
-    },
+        return this.websiteService.context.isMobile;
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.editor.js:416]: I don't think this is an equivalent change, right? I have been meaning to unify those. Maybe this deserves its own commit?

File: addons/website/static/src/js/editor/snippets.editor.js [hunk 15]
```diff
@@ -449,47 +486,61 @@ const wSnippetMenu = weSnippetEditor.SnippetsMenu.extend({
...
             }
+            return false;
         }
-    },
+    }
     /**
+     * Used to adjust the highlight effect when the text content is edited.
+     *
+     * TODO: Should be directly replaced by `switchTextHighlight()` in master
+     * (left in stable for compatibility).
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.editor.js:507]: Looks like a rebase fail on top of the recently merged commit of Mehdi. Have a look :pray:

File: addons/website/static/tests/tours/snippet_version.js [hunk 16]
```diff
@@ -24,25 +24,20 @@ wTourUtils.registerWebsitePreviewTour("snippet_version", {
...
 },
     ...wTourUtils.clickOnSave(),
-    ...wTourUtils.clickOnEditAndWaitEditMode(),
+]);
+wTourUtils.registerWebsitePreviewTour("snippet_version_2", {
+    edition: true,
+    url: "/",
+    test: true,
+}, () => [
+
```
Comment (qsm-odoo) [inline addons/website/static/tests/tours/snippet_version.js:39]: Blank line to remove

File: addons/website/tests/test_ui.py [hunk 17]
```diff
@@ -292,7 +287,42 @@ def test_07_snippet_version(self):
                 </xpath>
             """,
         }])
-        self.start_tour(self.env['website'].get_client_action_url('/'), 'snippet_version', login='admin')
+        self.start_tour(self.env['website'].get_client_action_url('/'), 'snippet_version_1', login='admin')
```
Comment (qsm-odoo) [inline addons/website/tests/test_ui.py:290]: Why did you actually have to split the tour in two ?

File: addons/website/tests/test_ui.py [hunk 18]
```diff
@@ -174,11 +174,6 @@ def test_snippet_translation(self):
                 'Contact us': 'Contact us in Parseltongue'
             }
         })
-        self.env.ref('web_editor.snippets').update_field_translations('arch_db', {
-            fake_user_lang.code: {
-                'Save': 'Save in fu_GB',
-            }
-        })
```
Comment (qsm-odoo) [inline addons/website/tests/test_ui.py:181]: Why this change?

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
@detrouxdev It looks already way better with the few first commits to split the main work :+1: 

I am adding a review: some things to fix, some questions (especially, some comments about the props vs state vs env vs etc) and some nitpicking. Probably answer all the questions first :pray:

File: addons/mass_mailing/static/src/js/mass_mailing_html_field.js [hunk 19]
```diff
@@ -25,10 +25,18 @@ export class MassMailingHtmlField extends HtmlField {
 
     setup() {
         super.setup();
+        // Use reactive instead of useState because currently, this component
+        // does not need to re-render when its state changes. If it does, change
+        // to `useState`.
+        this.state = reactive({
+            selectedTheme: null,
+            $scrollable: null,
+        });
```
[the remark this answers is not in the document]
Comment (detrouxdev) [reply addons/mass_mailing/static/src/js/mass_mailing_html_field.js:34]: > The reason it is not a state transferred as a prop is because you want to give it to a sub-sub-child, right? Or is there another reason?

Yes, that's the reason, to not pollute the wysiwyg props with a prop it will not use.

> The only code using this reactive object is addons/mass_mailing/static/src/js/snippets.editor.js, right? Actually does it even need to be reactive here? Is not the use of this.mailingState = useState(this.env.mailingState); enough?

Yes and no, only snippets.editor.js will read the variable, but it needs to be reactive before that so that setting something in that object triggers the reactive chain (re-render => useEffect)

In your example, doing `this.envConfig.selectedTheme = newTheme` would do nothing, as that object is a plain object, and not the reactive proxy. (Whenever an object is already reactive, using "useState" will subscribe the component's render method to the reactive object, it will not re-create a reactive object, so SnippetsMenu and MassMailingHtmlField share the same reactive object and can trigger stuff that way.)

> I am confused by this: "component does not need to re-render so no useState"... is it worth a comment? Should not you just not call it state?

I had a feeling that it could be confusing, that is why I tried to comment: I called it state because it's a reactive value that will cause a child component to re-render. I do not use `useState` because, from what I understand, setting a value in a state set by useState will cause OWL to re-render the component (I think at least once, but I might be wrong), which is not necessary here as the value is not really used by any OWL feature.

I do not mind calling it something else, but it would be nice to have a name that indicates that child components (only SnippetsMenu but any child component can subscribe to that reactive value) will react if that thing is accessed.

I'm okay with fieldConfig or fieldState, whichever you prefer.

File: addons/mass_mailing/static/src/js/snippets.editor.js [hunk 20]
```diff
@@ -29,10 +28,23 @@ export class MassMailingSnippetsMenu extends snippetsEditor.SnippetsMenu {
...
+        let firstRender = true;
+        useEffect(
+            (selectedTheme) => {
+                // Avoid running this on first render as the template is already
+                // loaded for the selected theme by the normal flow.
+                if (firstRender) {
+                    firstRender = false;
+                    return;
+                }
+                this._loadSnippetsTemplates().then(() => {
```
[the remark this answers is not in the document]
Comment (detrouxdev) [reply addons/mass_mailing/static/src/js/snippets.editor.js:38]: _loadSnippetsTemplate executes itself in the mutex, so I thought it would most likely be okay. I cannot await it in the useEffect (useEffect does not take async methods).

The only async methods that can be used in the lifecycle of an OWL Component are `onWillStart` and `onWillUpdateProps` but here, since the value is in the env, I did not use `onWillUpdateProps`. I guess that could be good reason enough to not use the env, and instead pass the variable through the props, even if wysiwyg does not use it. I just did not think there was a big risk since it's done through the mutex.

File: addons/website/static/src/components/wysiwyg_adapter/toolbar_patch.js (hunk 12 above)
Comment (detrouxdev) [reply addons/website/static/src/components/wysiwyg_adapter/toolbar_patch.js:15]: Only reason those were needed is because we were patching the `web_editor.toolbar` template to add those buttons. So we needed a way to hide them outside of website.

Now the buttons are defined in the [WebsiteSnippetsMenu template](https://github.com/odoo-dev/odoo/blob/b39c7a71881096de2beee1f336c66e5a455a9a2d/addons/website/static/src/xml/website.editor.xml#L196) so there is no need for that logic anymore.

File: addons/website/static/src/js/editor/snippets.editor.js (hunk 14 above)
Comment (detrouxdev) [reply addons/website/static/src/js/editor/snippets.editor.js:416]: wUtils.isMobile calls `service_context_get` with a trigger_up (which needed to be removed) [1](https://github.com/odoo/odoo/blob/b7bf4cc2272f8e004e6fa2fa1f7f11c9d8e9ba8e/addons/website/static/src/js/utils.js#L408)

that trigger_up is handled by `WysiwygAdapter` and returns `{ "isMobile": this.websiteService.context.isMobile }` [2](https://github.com/odoo/odoo/blob/b7bf4cc2272f8e004e6fa2fa1f7f11c9d8e9ba8e/addons/website/static/src/components/wysiwyg_adapter/wysiwyg_adapter.js#L1165)

Which I actually forgot to remove as I don't think the wysiwyg_adapter needs to handle it anymore.

File: addons/website/tests/test_ui.py (hunk 18 above)
Comment (detrouxdev) [reply addons/website/tests/test_ui.py:181]: I don't know how to change the translation of JS templates in Python. I tried looking for a way, but there is no models for it, it looks like [it is read directly from the disc](https://github.com/odoo/odoo/blob/b7bf4cc2272f8e004e6fa2fa1f7f11c9d8e9ba8e/odoo/tools/translate.py#L1660).

Comment (detrouxdev) [reply addons/website/tests/test_ui.py:181]: I guess I could patch the controller which fetches the translations, but I thought the JS way where I insert the translated term prior to running the snippets menu was a good enough replacement.

File: addons/mass_mailing/static/src/js/mass_mailing_html_field.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_html_field.js:34]: > In your example, doing this.envConfig.selectedTheme = newTheme would do nothing, as that object is a plain object, and not the reactive proxy.

Oh, ok, I see. I imagined that `reactive(obj) == obj`, makes sense that's not true.

> by useState will cause OWL to re-render the component (I think at least once, but I might be wrong)

I would imagine that's not true if the state is not used by the component XML definition. But not using `useState` seems good here given the fact it's a reactive stuff passed through env.

> I do not mind calling it something else

Not sure. I'll ask an OWL expert. I prepared a schema to do so:
image

[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_html_field.js:34]: EDIT, réponse de GED:
> en général, j'aime bien un nom qui représente le concept. du genre notification et pas notificationState. ou overlay, ou model

So I would avoid `state`, let's go with `fieldConfig`. `fieldReactiveConfig` if you prefer maybe :shrug:

File: addons/mass_mailing/static/src/js/snippets.editor.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/mass_mailing/static/src/js/snippets.editor.js:38]: If it's in the mutex, should be ok I guess, let's forget about this now :wink:

File: addons/website/tests/test_ui.py (hunk 18 above)
Comment (qsm-odoo) [reply addons/website/tests/test_ui.py:181]: Mmh indeed, probably ok :+1: Let's ping the test master to see if he has some insight @rdeodoo :open_mouth:

Comment (rdeodoo) [reply addons/website/tests/test_ui.py:181]: Sounds good to me, I also tested functionally and it still works (having the snippets body in the website lang regardless of the user / right panel lang)

File: addons/website/tests/test_ui.py (hunk 17 above)
Comment (detrouxdev) [reply addons/website/tests/test_ui.py:290]: Previously, the tour dropped the snippets, then accessed the html of the SnippetsMenu to increase the version of a snippet template, then clicked on the snippets to see if the "outdated" warning showed up.

This is not possible anymore since snippets are defined in SnippetsMenu as objects and are no longer in the DOM. So I split the tour in two, and use python to increase the version instead of using JS.

File: addons/mass_mailing/static/src/js/mass_mailing_wysiwyg.js [hunk 21]
```diff
@@ -104,7 +101,7 @@ export class MassMailingWysiwyg extends Wysiwyg {
         const res = await super._insertSnippetMenu();
         // Hide the snippetsMenu at first, other code will handle
         // if it should be shown or not.
-        this.snippetsMenu.setFolded(true);
+        this.snippetsMenuProps.folded = false;
```
[the remark this answers is not in the document]
Comment (detrouxdev) [reply addons/mass_mailing/static/src/js/mass_mailing_wysiwyg.js]: oops, no it isn't. I fixed it!

File: addons/web_editor/static/src/js/editor/snippets.editor.js [hunk 22]
```diff
@@ -1784,38 +1772,65 @@ var SnippetsMenu = Widget.extend({
...
+        onSnippetDropped: { type: Function, optional: true },
+        folded: { type: Boolean, optional: true },
+        "*": { optional: true },
+    };
+
+    static defaultProps = {
+        folded: false,
+        onSnippetDropped: () => {},
+        readyToCleanForSave: () => {},
+    };
```
[the remark this answers is not in the document]
Comment (detrouxdev) [reply addons/web_editor/static/src/js/editor/snippets.editor.js]: > What's the point of "*": { optional: true },

It allows passing any prop and ignore the props definition, thought `optional: true` is most likely not necessary

I removed it as it was a leftover of when I was trying to merge props and options. I also re-ordered props, and defined the missing ones. should be cleaner now 👍🏻

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 23]
```diff
@@ -663,11 +669,11 @@ export class Wysiwyg extends Component {
                     // editor panel (like originally intended but...) / ...
                     (async () => {
                         let container;
-                        if (this.snippetsMenu) {
+                        if (this.snippetsMenuBus) {
                             // Await for the editor panel to be fully updated
                             // as some buttons of the link popover we create
                             // here relies on clicking in that editor panel...
-                            await this.snippetsMenu._mutex.exec(() => null);
+                            await this.mutex.getUnlockedDef();
```
[the remark this answers is not in the document]
Comment (detrouxdev) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js]: You're right, though I'm not sure if this should not actually be getUnlockedDef as maybe we do want to wait  for everything even things that happens after this thing is called, before showing the LinkPopover.

For now, I restored the original way. 👍🏻

**Comment by qsm-odoo** [comment]
@odoo/rd-security Could you override the ci/security here please? :pray: 

> odoo/addons/mass_mailing/static/src/js/snippets.editor.js:216 preview: markup(mailingHtml.body.innerHTML),
> - JS004: markup( => 📚 View explanation

It's line that was already there and was moved (mass mailing code). FYI that "preview" value is used that way:
```javascript
   appendPreview() {
        const iframe = this.iframeRef.el.contentDocument;
        const body = iframe.querySelector("body");
        const isMarkup = this.props.preview instanceof markup().constructor;
        const safePreview = isMarkup ? this.props.preview : escape(this.props.preview);
        body.innerHTML = safePreview;
    }
```
There is clearly better to do (using the same system as the website one to just resize the current iframe to display the mobile preview instead of copying it), but that would be for another task and probably team :+1: 

Thanks in advance :slightly_smiling_face:

**Review by qsm-odoo — APPROVED** [review]
Thanks for the work @detrouxdev, a nice first step towards a clean OWL editor :muscle: 

I tested myself quite a bit, did not find anything relevant to say except one thing: in mass mailing, while *not* in fullscreen, the snippet's menu height is quite capricious... but it was already bad before: the html field keeps jumping in height depending on what is clicked. It should probably be fixed in stable and then we could care about it here. But if you want to have a look after this is merged, it's good too.

I did not find anything else problematic. After all, this went through several india testing already :clap: Let's not delay further, this is also why we wanted to merge as soon as possible after a freeze.

I made a final push-force with:
- a rebase
- some nitpicking (including what we talked about IRL)
- some commit message typos fixing (+ you forgot to adapt, it still spoke about `<category>` etc)

@robodoo rebase-ff r+ :partying_face:
